lib/srv/app/aws/assume_role.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/teleport/lib/srv/app/common"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+func updateAssumeRoleDuration(identity *tlsca.Identity, w http.ResponseWriter, req *http.Request, clock clockwork.Clock) error {
+	// Skip non-AssumeRole request
+	query, err := getAssumeRoleQuery(req)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			return nil
+		}
+		return trace.Wrap(err)
+	}
+
+	requestedDuration := getAssumeRoleDuration(query)
+	identityDuration := identity.Expires.Sub(clock.Now()).Round(time.Second)
+	switch {
+	// Deny access if identity duration is shorter than the minimum that can be
+	// requested.
+	case identityDuration < assumeRoleMinDuration:
+		return trace.AccessDenied("minimum AWS session duration is %v but Teleport identity expires in %v", assumeRoleMinDuration, identityDuration)
+
+	// Use shorter identity duration.
+	case identityDuration < requestedDuration:
+		setAssumeRoleDuration(query, identityDuration)
+		newBody := io.NopCloser(strings.NewReader(query.Encode()))
+		if err := utils.ReplaceRequestBody(req, newBody); err != nil {
+			return trace.Wrap(err)
+		}
+		w.Header().Add(common.TeleportAPIInfoHeader, fmt.Sprintf("requested DurationSeconds of AssumeRole is overwritten to \"%d\" as the Teleport identity will expire at %v", int(identityDuration.Seconds()), identity.Expires))
+		return nil
+
+	// Use shorter requested duration (no update required).
+	default:
+		return nil
+	}
```

Comment:
can you move this to a separate function ?

Reply:
I found it difficult to move the whole switch to a separate function. End up removing the switch and added a helper function. I think it should be easier to read now.